cmd/ova-parser/ova-parser.go

```diff
@@ -0,0 +1,336 @@
+	Disks   []Disk   `xml:"Disk"`
+}
+
+type Disk struct {
+	XMLName                 xml.Name `xml:"Disk"`
+	Capacity                string   `xml:"capacity,attr"`
+	CapacityAllocationUnits string   `xml:"capacityAllocationUnits,attr"`
+	DiskId                  string   `xml:"diskId,attr"`
+	FileRef                 string   `xml:"fileRef,attr"`
+	Format                  string   `xml:"format,attr"`
+	PopulatedSize           string   `xml:"populatedSize,attr"`
+}
+
+type NetworkSection struct {
+	XMLName  xml.Name  `xml:"NetworkSection"`
+	Info     string    `xml:"Info"`
+	Networks []Network `xml:"Network"`
+}
+
+type Network struct {
+	XMLName     xml.Name `xml:"Network"`
+	Name        string   `xml:"name,attr"`
+	Description string   `xml:"Description"`
+}
+
+type VirtualSystem struct {
+	ID                     string `xml:"id,attr"`
+	Name                   string `xml:"Name"`
+	OperatingSystemSection struct {
+		Info        string `xml:"Info"`
+		Description string `xml:"Description"`
+	} `xml:"OperatingSystemSection"`
+	HardwareSection VirtualHardwareSection `xml:"VirtualHardwareSection"`
+}
+
+type Envelope struct {
+	XMLName        xml.Name        `xml:"Envelope"`
+	VirtualSystem  []VirtualSystem `xml:"VirtualSystem"`
+	DiskSection    DiskSection     `xml:"DiskSection"`
+	NetworkSection NetworkSection  `xml:"NetworkSection"`
+}
+
+// vm struct
+type VM struct {
+	Name                  string
+	OvaPath               string
+	RevisionValidated     int64
+	PolicyVersion         int
+	UUID                  string
+	Firmware              string
+	CpuAffinity           []int32
+	CpuHotAddEnabled      bool
+	CpuHotRemoveEnabled   bool
+	MemoryHotAddEnabled   bool
+	FaultToleranceEnabled bool
+	CpuCount              int32
+	CoresPerSocket        int32
+	MemoryMB              int32
+	BalloonedMemory       int32
+	IpAddress             []string
+	NumaNodeAffinity      []string
+	StorageUsed           int64
+	ChangeTrackingEnabled bool
+	Devices               []Device
+	NICs                  []NIC
+	Disks                 []VmDisk
+	Networks              []Network
+}
+
+// Virtual Disk.
+type VmDisk struct {
+	FilePath                string
+	Capacity                string
+	CapacityAllocationUnits string
+	DiskId                  string
+	FileRef                 string
+	Format                  string
+	PopulatedSize           string
+}
+
+// Virtual Device.
+type Device struct {
+	Kind string `json:"kind"`
+}
+
+type Conf struct {
+	key   string
+	Value string
+}
+
+// Virtual ethernet card.
+type NIC struct {
+	Name   string `json:"name"`
+	MAC    string `json:"mac"`
+	Config []Conf
+}
+
+func main() {
+	ovaFiles, err := findOVAFiles("/ova")
+	if err != nil {
+		fmt.Println("Error finding OVA files:", err)
+		return
+	}
+
+	for _, ovaFile := range ovaFiles {
+		fmt.Println("Processing OVA file:", ovaFile)
+
+		ovfPath, tmpDir, err := extractOVFFromOVA(ovaFile)
+		if err != nil {
+			fmt.Println("Error extracting OVF from OVA:", err)
+			continue
+		}
```

Comment:
I suggest using `defer os.RemoveAll(tmpDir)` here instead calling it separately below. Either way you may want to log the error from `os.RemoveAll`.

Reply:
Using a library is probably preferable if it can be used easily. But you can also tell `tar` to dump the file to stdout with `-O/--to-stdout`. But you also need to call `tar` multiple times -- first to list the files and then get the output. (Just saying.)

Reply:
> But you also need to call `tar` multiple times -- first to list the files and then get the output. (Just saying.)

according to the OVF specification, the OVF descriptor should always be the first entry ("The ordering restriction ensures that it is possible to extract the OVF descriptor from a compressed OVF package without scanning the entire archive. ")

Reply:
In theory yes. In reality there are plenty of hand crafted OVAs not following the specs. Some of which are not even TAR files (are e.g. ZIP files). And it's fine to discard those as not supported. But you don't want such files to cause you troubles.

Specifically in this context, if you just run `tar xOf somethig.ova --wildcards '*.ovf'` and the archive contains multiple OVF files you will end up with their content concatenated together which will confuse the XML parser and produce bogus errors. That's what I meant.